File: Implementation/Networking/Minimum_Spanning_Tree/prim.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <climits>
#include <algorithm>

using namespace std;

struct arc {
    int src;
    int dest;
    int weight;

    arc(int src, int dest, int weight): src(src), dest(dest), weight(weight) {
    }
};

struct dist {
    int node;
    int distance;

    dist(int node, int distance) : node(node), distance(distance) {
    }
};

class CompareDist {
public:
    bool operator()(const dist* a, const dist* b) const {
        return a->distance > b->distance;
    }
};

int prim(const vector<arc> graph, vector<arc> &mst, int node_num) {
    int total = 0, node_idx, update_node;
    vector<dist*> dist_mst;
    int *pre = new int[node_num];
    dist **node_list = new dist*[node_num];

    for (int i=0; i < node_num; i++) {
        node_list[i] = new dist(i, INT_MAX);
        dist_mst.push_back(node_list[i]);
    }
    node_list[0]->distance = 0;
    fill_n(pre, node_num, -1);

    while (!dist_mst.empty()) {
        dist* node = dist_mst.front();
        total += node->distance;
        node_idx = node->node;
        mst.emplace_back(pre[node_idx], node_idx, node->distance);
        for (arc a : graph) {
            if (a.src == node_idx || a.dest == node_idx) {
                update_node = a.src + a.dest - node_idx;
                if (a.weight < node_list[update_node]->distance) {
                    pre[update_node] = node_idx;
                    node_list[update_node]->distance = a.weight;
                }
            }
        }
        dist_mst.erase(dist_mst.begin());
        make_heap(dist_mst.begin(), dist_mst.end(), CompareDist());
    }

    return total;
}

int main(int argc, char** argv) {
    ifstream in(argv[1]);
    vector<arc> graph;
    vector<arc> mst;

    int node_num, arc_num, dummy;
    int src, dest, weight;

    in >> node_num >> arc_num >> dummy;
    while (in >> src >> dest >> weight)
        graph.emplace_back(src, dest, weight);

    int total = prim(graph, mst, node_num);

    cout << "Total weight: " << total << endl;
    for (arc a : mst)
        cout << "{" << a.src << ", " << a.dest << ", " << a.weight << "}";

    cout << endl;

    return 0;
}

File: Implementation/Networking/Minimum_Spanning_Tree/prim.rs
extern crate priority_queue;

use std::env;
use std::fs;
use std::u32;
use std::cmp::Reverse;
use priority_queue::PriorityQueue;

struct Arc {
    src: usize,
    dest: usize,
    weight: u32,
}

fn prim(graph: &Vec<Arc>, mst: &mut Vec<Arc>, node_num: usize) -> u32 {
    let mut total = 0;
    let mut dist_mst = PriorityQueue::new();
    let mut pre = vec![0; node_num];

    for x in 0..node_num {
        dist_mst.push(x, Reverse(u32::max_value()));
    }
    dist_mst.change_priority(&0, Reverse(0));
    
    while !dist_mst.is_empty() {
        let node = dist_mst.pop().unwrap();
        let node_idx = node.0;
        let priority = node.1;
        total += priority.0;
        let mst_node = Arc {src : pre[node_idx], dest : node_idx, weight : priority.0};
        mst.push(mst_node);

        for arc in graph.iter() {
            if arc.src == node_idx || arc.dest == node_idx {
                let update_node = arc.src + arc.dest - node_idx;
                let p = dist_mst.get_priority(&update_node).cloned();
                match p {
                    Some(priority) => {
                        if arc.weight < priority.0 {
                            pre[update_node] = node_idx;
                            dist_mst.change_priority(&update_node, Reverse(arc.weight));
                        }
                    }
                    None => {
                    }
                }
            }
        }
    }

    total
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let filename = &args[1];
    let mut graph: Vec<Arc> = Vec::new();
    let mut mst: Vec<Arc> = Vec::new();

    let f = fs::read_to_string(filename).expect("file not found");
    let mut file: Vec<_> = f.lines().collect();
    let info: Vec<_> = file[0].split_whitespace().collect();
    let node_num = info[0].parse::<usize>().unwrap();
    file.remove(0);

    for line in file {
        let info: Vec<_> = line.split_whitespace().collect();
        let new_arc = Arc {
            src: info[0].parse::<usize>().unwrap(),
            dest: info[1].parse::<usize>().unwrap(),
            weight: info[2].parse::<u32>().unwrap(),
        };
        graph.push(new_arc);
    }

    let total = prim(&graph,&mut mst, node_num);

    println!("MST with total weight: {}", total);
    for arc in mst.iter() {
        println!("({}, {}, {})", arc.src, arc.dest, arc.weight);
    }
}
